### PR TITLE
[Merged by Bors] - Handle updated VRF Nonces correctly without VRFSigner service

### LINKED ISSUE #3902: Re-assess nonce encoding in VRF signatures
## Description
With https://github.com/spacemeshos/go-spacemesh/pull/3877 nonces were added to VRF signatures. These are currently just added as an 8 byte little endian prefix to the message that is signed before signing (and validation).

We should re-assess how we want to encode this nonce (see also https://github.com/spacemeshos/go-spacemesh/pull/3877#discussion_r1056026113) :
- Add it to the message as a field before serializing it with go-scale
- using a separate `VRFMessage` type that has 2 fields: the message to be signed and the nonce before encoding that with go-scale
- possibly a different encoding

Both have advantages and disadvantages; using it as a field in the actual message means we can have messages with and without the nonce, and the messages can be validated without first needing to fetch and validate the initial ATX of a node. The second requires less code change and keeps the message objects themselves clean to contain only the data that is relevant for the message itself.

## Affected code
`signing` package and possibly `common/types` for the objects that are signed with the VRFSigner

### LINKED ISSUE #3962: Allow nodes to update their VRF Nonce
### Description
Nodes must be able to update their VRF Nonce in case they increase their allocated PoST size.

### Acceptance criteria
- Any incoming ATX (instead of just initial ATXs) may contain a VRFNonce that shall be used for the identity starting in the `TargetEpoch` of the ATX.
- New VRFNonces need to be verified to satisfy the required difficulty as set by the claimed PoST size.

### Implementation hints
- Instead of fetching the VRF Nonce from the initial ATX of an identity, VRF Nonces should be stored along with the identity they belong to and the epoch from which on they become valid in the nodes database
   - When verifying a VRF signature the epoch in which it was produced needs to be used to fetch the right VRF Nonce for the given identity
   - previous VRF Nonces should not be removed from the local node state until some cohesive strategy for pruning nodes has been specified

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Alternative approach for https://github.com/spacemeshos/go-spacemesh/pull/3966

Closes https://github.com/spacemeshos/go-spacemesh/issues/3902 and closes https://github.com/spacemeshos/go-spacemesh/issues/3962

## Changes
Instead of using a service to fetch nonces and sign VRF messages, put the logic into the services themselves and make the VRF Signer/Verifier only sign messages

## Test Plan
n/a

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by fasmat** [comment]
bors try

File: signing/signer.go [hunk 1]
```diff
@@ -114,22 +115,9 @@ func (es *EdSigner) PrivateKey() PrivateKey {
...
-	}
-
-	if err := cfg.validate(); err != nil {
-		return nil, err
-	}
-
+func (es *EdSigner) VRFSigner() (*VRFSigner, error) {
 	return &VRFSigner{
-		privateKey: es.priv,
+		privateKey: oasis.PrivateKey(es.priv),
```
Comment (countvonzero) [inline signing/signer.go:120]: i am confused by this cast. why is this needed?

File: proposals/handler_test.go [hunk 2]
```diff
@@ -61,21 +61,22 @@ type testHandler struct {
...
+		mNonce: NewMocknonceFetcher(ctrl),
 	}
 }
 
 func createTestHandler(t *testing.T) *testHandler {
 	types.SetLayersPerEpoch(layersPerEpoch)
 	ms := fullMockSet(t)
 	return &testHandler{
-		Handler: NewHandler(datastore.NewCachedDB(sql.InMemory(), logtest.New(t)), ms.mpub, ms.mf, ms.mbc, ms.mm, ms.md,
+		Handler: NewHandler(datastore.NewCachedDB(sql.InMemory(), logtest.New(t)), ms.mpub, ms.mf, ms.mbc, ms.mm, ms.md, ms.mvrf, ms.mNonce,
```
Comment (countvonzero) [inline proposals/handler_test.go]: looks like its not needed in this test.

File: beacon/message.go [hunk 3]
```diff
@@ -14,6 +14,7 @@ import (
 type ProposalMessage struct {
 	EpochID      types.EpochID
 	NodeID       types.NodeID
+	Nonce        types.VRFPostIndex
```
Comment (countvonzero) [inline beacon/message.go]: this is incorrect. we don't need Nonce here. this is not the VRF message being signed. this is just the msg sent over the wire during the beacon protocol.

File: beacon/weakcoin/weak_coin.go [hunk 4]
```diff
@@ -222,6 +235,7 @@ func (wc *WeakCoin) prepareProposal(epoch types.EpochID, round types.RoundID) ([
 				Round:     round,
 				Unit:      unit,
 				MinerPK:   wc.signer.PublicKey().Bytes(),
+				Nonce:     nonce,
```
Comment (countvonzero) [inline beacon/weakcoin/weak_coin.go]: same here. this is the gossiped message. not the vrf message. don't add nonce here.

**Review by countvonzero — COMMENTED** [review]
you can also close https://github.com/spacemeshos/go-spacemesh/issues/3902 with this PR

File: signing/signer.go (hunk 1 above)
Comment (fasmat) [reply signing/signer.go:120]: `es.priv` is a `crypto.PrivateKey` defined in the standard library. The but the signer uses `"github.com/oasisprotocol/curve25519-voi/primitives/ed25519"` which for some reason defined its own type of `PrivateKey`. Both are just `[]byte` but there are no implicit casts in go (at least not from non-basic type to non-basic type), so this explicit cast is necessary.

File: proposals/handler_test.go (hunk 2 above)
Comment (fasmat) [reply proposals/handler_test.go]: Not in this file, but `createTestHandler` is also used in `elegibility_validator_test.go` and there I set expectations on those mocks.

File: cmd/node/node.go [hunk 5]
```diff
@@ -458,6 +459,14 @@ func (app *App) SetLogLevel(name, loglevel string) error {
 	return nil
 }
 
+type nonceFetcherAdapter struct {
+	cdb *datastore.CachedDB
+}
+
+func (nfa nonceFetcherAdapter) VRFNonce(nodeID types.NodeID, epoch types.EpochID) (types.VRFPostIndex, error) {
+	return atxs.VRFNonce(nfa.cdb, nodeID, epoch)
+}
```
Comment (fasmat) [inline cmd/node/node.go]: This is probably not the right place for this; but I'm not sure in which package to put it? `datastore`?

Comment (countvonzero) [reply cmd/node/node.go]: i think i will prefer creating a option for each component's `NewXXXX` function.
that option will default to a function that executes `txs.VRFNonce(cdb, nodeID, epoch)` (because cdb is already passed to most components), and gets overwritten in tests

Comment (countvonzero) [reply cmd/node/node.go]: similar to this https://github.com/spacemeshos/go-spacemesh/blob/3c2f1837675840e69e3f5bb5cabbd9a9e54807ce/proposals/handler.go#L79

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try-

**Comment by fasmat** [comment]
bors -try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

File: cmd/node/node.go (hunk 5 above)
Comment (fasmat) [reply cmd/node/node.go]: I added a few functional options to inject a mocked `nonceFetcher` in tests. If not used a standard implementation is used that reads the nonce from the ATX table using `sql/atxs`

File: beacon/beacon.go [hunk 6]
```diff
@@ -669,16 +688,22 @@ func (pd *ProtocolDriver) sendProposal(ctx context.Context, epoch types.EpochID)
 	}
 
 	logger := pd.logger.WithContext(ctx).WithFields(epoch)
-	proposedSignature := buildSignedProposal(ctx, pd.vrfSigner, epoch, pd.logger)
+	nonce, err := pd.nonceFetcher.VRFNonce(pd.nodeID, epoch)
+	if err != nil {
+		logger.With().Panic("failed to get VRF nonce", log.Err(err))
```
Comment (countvonzero) [inline beacon/beacon.go]: prefer `logger.Fatal()`
stack trace is usually useless.

File: beacon/weakcoin/weak_coin.go [hunk 7]
```diff
@@ -78,17 +79,23 @@ func WithNextRoundBufferSize(size int) OptionFunc {
...
 	wc := &WeakCoin{
-		logger:    log.NewNop(),
-		config:    defaultConfig(),
-		signer:    signer,
-		publisher: publisher,
-		coins:     make(map[types.RoundID]bool),
-		verifier:  verifier,
+		logger:       log.NewNop(),
+		config:       defaultConfig(),
+		cdb:          cdb,
```
Comment (countvonzero) [inline beacon/weakcoin/weak_coin.go]: not needed. the nonceFetcher is sufficient.

File: beacon/weakcoin/weak_coin_test.go [hunk 8]
```diff
@@ -337,6 +364,12 @@ func TestWeakCoinNextRoundBufferOverflow(t *testing.T) {
 	require.True(t, wc.Get(context.Background(), epoch, nextRound))
 }
 
+type vrfVerifierAdapter struct{}
+
+func (m vrfVerifierAdapter) Verify(nodeID types.NodeID, msg, sig []byte) bool {
```
Comment (countvonzero) [inline beacon/weakcoin/weak_coin_test.go]: why is this needed?

File: beacon/weakcoin/weak_coin_test.go [hunk 9]
```diff
@@ -358,19 +391,25 @@ func TestWeakCoinEncodingRegression(t *testing.T) {
...
+		types.RandomNodeID(),
+		vrfSig,
+		vrfVerifierAdapter{},
+		nonceFetcher(t, ctrl),
+		weakcoin.WithThreshold([]byte{0xff}),
+	)
 	instance.StartEpoch(context.Background(), epoch, allowances)
 	require.NoError(t, instance.StartRound(context.Background(), round))
 
+	// TODO(mafa): figure out why sig is now different on every test run
```
Comment (countvonzero) [inline beacon/weakcoin/weak_coin_test.go]: but it should change. nonce is an extra data to the vrf message now.

**Review by countvonzero — APPROVED** [review]

File: beacon/weakcoin/weak_coin_test.go (hunk 8 above)
Comment (fasmat) [reply beacon/weakcoin/weak_coin_test.go]: When I wrote that code there VRFVerifier was just a function so I needed to wrap it into an object to satisfy the interface. Since then I created a `VRFVerifier` type in the `signing` package and this is not needed any more.

Removed.

File: beacon/weakcoin/weak_coin_test.go (hunk 9 above)
Comment (fasmat) [reply beacon/weakcoin/weak_coin_test.go]: The problem was that every time I ran the test the signature would be different (without changing code). I've fixed the issue so this comment is obsolete.

Removed.

File: sql/atxs/atxs.go [hunk 10]
```diff
@@ -208,6 +208,30 @@ func GetIDsByEpoch(db sql.Executor, epoch types.EpochID) (ids []types.ATXID, err
...
+		stmt.BindInt64(2, int64(epoch))
+	}
+	dec := func(stmt *sql.Statement) bool {
+		nonce = types.VRFPostIndex(stmt.ColumnInt64(0))
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select nonce from atxs
+		where smesher = ?1 and epoch <= ?2 and nonce is not null
```
Comment (countvonzero) [inline sql/atxs/atxs.go]: the recorded epoch is publish epoch. so here i think you want `<` instead of `<=`

Comment (countvonzero) [reply sql/atxs/atxs.go]: never mind. i am wrong.

Comment (countvonzero) [reply sql/atxs/atxs.go]: actually you do want `<` here. the epoch you passed in is target epoch

Comment (fasmat) [reply sql/atxs/atxs.go]: You are probably right: when I wrote that query I was persisting the nonce based on `TargetEpoch`, but `epoch` here refers to `PublishEpoch`. Changed `<=` to `<`.

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

File: hare/algorithm.go [hunk 11]
```diff
@@ -239,6 +240,7 @@ func newConsensusProcess(
 	stateQuerier stateQuerier,
 	signing Signer,
 	nid types.NodeID,
+	nonce types.VRFPostIndex,
```
Comment (countvonzero) [inline hare/algorithm.go:243]: this needs to be assigned.

File: hare/interfaces.go [hunk 12]
```diff
@@ -17,7 +17,7 @@ type layerPatrol interface {
 type Rolacle interface {
 	Validate(context.Context, types.LayerID, uint32, int, types.NodeID, []byte, uint16) (bool, error)
 	CalcEligibility(context.Context, types.LayerID, uint32, int, types.NodeID, []byte) (uint16, error)
```
Comment (countvonzero) [inline hare/interfaces.go]: this needs to add nonce too. this is only used by own hare generating proof.

File: hare/eligibility/oracle.go [hunk 13]
```diff
@@ -167,15 +185,16 @@ func (o *Oracle) getBeaconValue(ctx context.Context, epochID types.EpochID) (uin
...
-		log.Uint32("beacon_val", value))
+		log.Uint32("beacon_val", value),
+	)
 	return value, nil
 }
 
 // buildVRFMessage builds the VRF message used as input for the BLS (msg=Beacon##Layer##Round).
-func (o *Oracle) buildVRFMessage(ctx context.Context, layer types.LayerID, round uint32) ([]byte, error) {
+func (o *Oracle) buildVRFMessage(ctx context.Context, nonce types.VRFPostIndex, layer types.LayerID, round uint32) ([]byte, error) {
 	key := buildKey(layer, round)
```
Comment (countvonzero) [inline hare/eligibility/oracle.go]: need to get rid of the vrfMsgCache. this is what causing the vrf signature to be invalid. once the node sign its own vrf message, it is validating all other nodes' vrf signature with the cached msg

Comment (fasmat) [reply hare/eligibility/oracle.go]: I removed the cache and fixed all unit tests that broke because of it; running a new `bors try` now

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
bors merge
